src/ch/ch_hotplug.c

```diff
@@ -0,0 +1,194 @@
+ * License along with this library.  If not, see
+ * <http://www.gnu.org/licenses/>.
+ */
+
+#include "ch_cgroup.h"
+#include "ch_domain.h"
+#include "ch_hotplug.h"
+#include "ch_monitor.h"
+#include "ch_process.h"
+#include "domain_audit.h"
+
+#define VIR_FROM_THIS VIR_FROM_CH
+
+/**
+ * virCHDomainSelectHotplugVcpuEntities:
+ *
+ * @def: domain definition
+ * @nvcpus: target vcpu count
+ *
+ * Validate vcpu entities will be able to be enabled/disabled.
+ *
+ * Returns the true if the nvcpus can be set, false if not.
+ */
+static bool
+virCHDomainSelectHotplugVcpuEntities(virDomainDefPtr def,
+                                     unsigned int nvcpus)
+{
+    bool ret = false;
+    virDomainVcpuDefPtr vcpu;
+    unsigned int maxvcpus = virDomainDefGetVcpusMax(def);
+    unsigned int curvcpus = virDomainDefGetVcpus(def);
+    ssize_t i;
+
+    if (nvcpus > curvcpus) {
+        for (i = 0; i < maxvcpus && curvcpus < nvcpus; i++) {
+            vcpu = virDomainDefGetVcpu(def, i);
+
+            if (vcpu->online)
+                continue;
+
+            curvcpus++;
+
+            if (curvcpus > nvcpus) {
+                virReportError(VIR_ERR_CONFIG_UNSUPPORTED, "%s",
+                               _("target vm vcpu granularity does not allow the "
+                                 "desired vcpu count"));
+                goto error;
+            }
+        }
+    } else {
+        for (i = maxvcpus - 1; i >= 0 && curvcpus > nvcpus; i--) {
+            vcpu = virDomainDefGetVcpu(def, i);
+
+            if (!vcpu->online)
+                continue;
+
+            curvcpus--;
+
+            if (curvcpus < nvcpus) {
+                virReportError(VIR_ERR_CONFIG_UNSUPPORTED, "%s",
+                               _("target vm vcpu granularity does not allow the "
+                                 "desired vcpu count"));
+                goto error;
+            }
+        }
+    }
+
+    if (curvcpus != nvcpus) {
+        virReportError(VIR_ERR_CONFIG_UNSUPPORTED, "%s",
+                       _("failed to find appropriate hotpluggable vcpus to "
+                         "reach the desired target vcpu count"));
+        goto error;
+    }
+
+    ret = true;
+
+ error:
+    return ret;
+}
+
+static int
+virCHDomainHotplugSetVcpus(virCHDriverPtr driver,
+                           virCHDriverConfigPtr cfg,
+                           virDomainObjPtr vm,
+                           unsigned int nvcpus)
+{
+    int oldvcpus = virDomainDefGetVcpus(vm->def);
+    virCHDomainObjPrivatePtr priv = vm->privateData;
+    virCHMonitorPtr mon = priv->monitor;
+    int ret = -1;
+
+    ret = virCHMonitorResizeCPU(mon, nvcpus);
+
+    virDomainAuditVcpu(vm, oldvcpus, nvcpus, "update", ret == 0);
+
+    if (ret < 0)
+        goto error;
+
+    for (int i = 0; i < nvcpus; i++) {
+        virDomainVcpuDefPtr vcpuinfo = virDomainDefGetVcpu(vm->def, i);
+        vcpuinfo->online = true;
+    }
+
+    if (virCHMonitorRefreshThreadInfo(mon) < 0)
```

Comment:
I feel that we need not have this logic here. When there is a change in the vcpu threads, we might get a monitor event and the monitor thread might take care of it. Could you please give it a  try?